singularity-eos/base/fast-math/logs.hpp

```diff
@@ -78,76 +62,166 @@
 namespace singularity {
 namespace FastMath {
 
-// Integer Aliased versions. Likely not as portable
-// TODO(JMM): Add single-precision versions
+template <typename T>
+PORTABLE_FORCEINLINE_FUNCTION auto sgn(const T &x) {
+  return (T(0) < x) - (x < T(0));
+}
+
 PORTABLE_FORCEINLINE_FUNCTION
-auto as_long_int(double f) { return *reinterpret_cast<long long int *>(&f); }
+auto as_int(double f) { return *reinterpret_cast<std::int64_t *>(&f); }
 PORTABLE_FORCEINLINE_FUNCTION
-auto as_double(long long int i) { return *reinterpret_cast<double *>(&i); }
+auto as_double(std::int64_t i) { return *reinterpret_cast<double *>(&i); }
```

Comment:
Can these (and subsequent) functions be `constexpr`?

Reply:
Huh. Led me to `std::bit_cast` (https://en.cppreference.com/w/cpp/numeric/bit_cast). It's C++20 but does have the tempting "Possible Implementation"

```c++
template<class To, class From>
std::enable_if_t<
    sizeof(To) == sizeof(From) &&
    std::is_trivially_copyable_v<From> &&
    std::is_trivially_copyable_v<To>,
    To>
// constexpr support needs compiler magic
bit_cast(const From& src) noexcept
{
    static_assert(std::is_trivially_constructible_v<To>,
        "This implementation additionally requires "
        "destination type to be trivially constructible");
 
    To dst;
    std::memcpy(&dst, &src, sizeof(To));
    return dst;
}
```

I'm guessing ``// constexpr support needs compiler magic`` is the fun part here.

Reply:
lol that seems plausible. I think there's a gcc extension? At least that's what my googling led me to believe. I don't think we want to rely on that though.

Reply:
Ah, the issue is `std::memcpy` accepts `void*` which can't be `constexpr` because the compiler isn't allowed to evaluate `void*` to an explicit type! same seems to apply to `reinterpret_cast`.

Found [this](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0202r0.html) (see III.B) linked on SO.

Reply:
Ah, which explains why it's a compiler extension but not directly in the standard. Good find!

Reply:
I like there being "Possible Implementations" on [cppreference](https://en.cppreference.com) and don't see why it shouldn't be used in most cases, but I don't think I've come across one that was had a "not really" comment attached.

Nah, shouldn't use it. Left at the mercy of C compatibility, yet again!

Reply:
Well, `std::variant` _is_ just a fancy `union`

Reply:
fair point.